consensus/src/dag/dag_fetcher.rs

```diff
@@ -327,19 +334,66 @@ impl TDagFetcher for DagFetcher {
+                        FetchResponse::try_from(response).and_then(|response| {
+                            response.verify(&remote_request_clone, &self.epoch_state.verifier)
+                        })
+                    }
+                    .and_then(|response| {
+                        let epoch_state = self.epoch_state.clone();
+                        let bounded_executor = bounded_executor.clone();
+                        async move {
+                            let nodes = response.certified_nodes();
+                            let epoch = epoch_state.epoch;
+                            ensure!(
+                                stream::iter(nodes.clone())
+                                    .concurrent_map(bounded_executor.clone(), move |node| {
+                                        let epoch_state = epoch_state.clone();
+                                        async move { node.verify(&epoch_state.verifier).is_ok() }
+                                    })
+                                    .all(|ok| async move { ok })
+                                    .await,
+                                "unable to verify certified nodes"
+                            );
+                            Ok(nodes)
+                        }
+                    })
+                    .await
+                    {
+                        Ok(certified_nodes) => {
                             // TODO: support chunk response or fallback to state sync
                             {
-                                for node in certified_nodes.into_iter().rev() {
-                                    if let Err(e) = dag.add_node(node) {
-                                        error!(error = ?e, "failed to add node");
+                                let groups: Vec<(Round, Vec<_>)> = certified_nodes
+                                    .into_iter()
+                                    .rev()
+                                    .group_by(|node| node.round())
+                                    .into_iter()
+                                    .map(|(key, group)| (key, group.collect()))
+                                    .collect();
+                                for (_round, round_group) in groups {
+                                    let mut tasks = FuturesUnordered::new();
+
+                                    for node in round_group.into_iter() {
+                                        let dag = dag.clone();
+                                        let handle = bounded_executor
+                                            .spawn(async move {
+                                                if let Err(e) = dag.add_node(node) {
+                                                    error!(error = ?e, "failed to add node");
+                                                }
+                                            })
+                                            .await;
+                                        tasks.push(handle);
```

Comment:
I "think" you need to wait previous task finish before creating new ones for next round otherwise there could be a race insertion fails with missing dependencies

Reply:
yes, that was my intent with grouping. fixed it.